### PR TITLE
Move file input operation to top level of filelogreceiver

### LINKED ISSUE #2265: Refactor stanzareceiver to allow splitting into separate receivers
We would like to separate the "source" of the log from the "operators" (it is currently one list in stanzareceiver). The goal is to be able to do this in the config:

```yaml
filelog:
 include: [ receiver/stanzareceiver/testdata/simple.log ]
 start_at: beginning
 operators:
   - type: regex_parser
       regex: '^(?P<time>\d{4}-\d{2}-\d{2}) (?P<sev>[A-Z]*) (?P<msg>.*)$'
       timestamp:
         parse_from: time
         layout: '%Y-%m-%d'
       severity:
         parse_from: sev
```
 
```yaml
winlog:
 channels: [Application, System]
 # other winlog options.
 operators:
   # similar to filelog.
```
 
```yaml
journald:
 # optional settings.
 directory:
 files:
 operators:
   # similar to filelog.
```

### LINKED ISSUE #2268: Add filelog receiver
Implement "filelog" receiver. The configuration should allow specifying the input and the list of operators (like stanzareceiver does).

An example config is this:
```yaml
filelog:
 include: [ receiver/stanzareceiver/testdata/simple.log ]
 start_at: beginning
 operators:
   - type: regex_parser
       regex: '^(?P<time>\d{4}-\d{2}-\d{2}) (?P<sev>[A-Z]*) (?P<msg>.*)$'
       timestamp:
         parse_from: time
         layout: '%Y-%m-%d'
       severity:
         parse_from: sev
```

### LINKED ISSUE #2282: Make sure regex, json, timestamp, severity parser operators are supported for all log receivers.
Sub-issue of #2264.

This will is likely to be solved as part of #2265.

### PR DESCRIPTION
**Pull request by djaglowski**
**Link to tracking Issue:**

Resolves #2265, Resolves #2268, and Resolves #2282.

**External Changes**

Moves the file input operator to the top of the`filelog` configuration:
```
receivers:
  filelog:
    include: [ testdata/simple.log ]
    start_at: beginning
    operators:
      - type: regex_parser
        regex: '^(?P<time>\d{4}-\d{2}-\d{2}) (?P<sev>[A-Z]*) (?P<msg>.*)$'
        timestamp:
          parse_from: time
          layout: '%Y-%m-%d'
        severity:
          parse_from: sev
```

---

Updates the `stanzareceiver.LogReceiverType`. It pulls more standard functionality into the `stanzareceiver` package, such as decoding the parser configs. However, I couldn't find a way to avoid adding a fairly "dumb" method, `BaseConfig(configmodels.Receiver) BaseConfig`, which just needs to grab `stanzareceiver.BaseConfig` from the individual implementation. I'm 99% sure there's a better way to do this, but I'm struggling to find it at the moment.

**Next Steps**

The `filelog` readme is badly out of date. Captured in #2335

---

The stanza operators still do not make use of `mapstructure`, which forces the configuration to work with generic `map[string]interface{}` and `map[interface{}]interface{}`. This is messy and makes it difficult to write good tests. I want to clean this up soon. Captured in several [issues](https://github.com/open-telemetry/opentelemetry-log-collection/issues) on the `opentelemetry-log-collection` repo.

**Testing:**
Merged `filelogreceiver/e2e_test.go` into `filelogreceiver/filelog_test.go`, but set of tests is the same.
Testbed still passes: 
```
=== RUN   TestLog10kDPS/filelog
2021/02/11 20:22:54 Starting mock backend...
... (abbreviated)
2021/02/11 20:23:10 Stopped backend. Received:   150,000 items (9,573/sec)
--- PASS: TestLog10kDPS (32.22s)
    --- PASS: TestLog10kDPS/OTLP (16.55s)
    --- PASS: TestLog10kDPS/filelog (15.67s)
PASS
ok      github.com/open-telemetry/opentelemetry-collector-contrib/testbed/tests_unstable_exe    32.855s
# Test PerformanceResults
Started: Thu, 11 Feb 2021 20:22:37 -0500

Test                                    |Result|Duration|CPU Avg%|CPU Max%|RAM Avg MiB|RAM Max MiB|Sent Items|Received Items|
----------------------------------------|------|-------:|-------:|-------:|----------:|----------:|---------:|-------------:|
Log10kDPS/OTLP                          |PASS  |     17s|    11.6|    12.7|         37|         46|    150000|        150000|
Log10kDPS/filelog                       |PASS  |     16s|    23.9|    25.9|         40|         48|    150000|        150000|
```

### CODE REVIEW

**Comment by djaglowski** [comment]
I'll add a few more tests to catch the error conditions that aren't covered.

File: receiver/filelogreceiver/README.md [hunk 1]
```diff
@@ -44,10 +44,9 @@ It can also be easily configured to tail and parse any structured or unstructure
 Receiver Configuration
 ```yaml
 receivers:
-  stanza:
-    operators:
-      - type: file_input
-        include: [ /var/log/myservice/*.json ]
+  filelog:
+    include: [ /var/log/myservice/*.json ]
+    parsers:
```
Comment (tigrannajaryan) [inline receiver/filelogreceiver/README.md:49]: >Renames operators to parsers. I believe this term conveys the intention of this sub-pipeline better than operators, even though we may end up supporting some operations that are not truly just "parsing". For example, we may want to be able add/delete/move fields.

If we do this will these non-parser operators be placed under "parsers" or we will have a different subkey to put them under?

File: receiver/filelogreceiver/filelog.go [hunk 2]
```diff
@@ -43,22 +42,45 @@ func (f ReceiverType) Type() configmodels.Type {
...
+func (f ReceiverType) BaseConfig(cfg configmodels.Receiver) stanzareceiver.BaseConfig {
+	return cfg.(*FileLogConfig).BaseConfig
 }
 
 // FileLogConfig defines configuration for the filelog receiver
 type FileLogConfig struct {
-	configmodels.ReceiverSettings `mapstructure:",squash"`
-	Operators                     stanzareceiver.OperatorConfig `mapstructure:"operators"`
+	stanzareceiver.BaseConfig `mapstructure:",squash"`
+	Input                     stanzareceiver.InputConfig `mapstructure:",remain"`
```
Comment (tigrannajaryan) [inline receiver/filelogreceiver/filelog.go:70]: So `mapstructure:",remain"` ensures that everything that is not declared by stanzareceiver.BaseConfig is captured in `Input`? Am I right?

One thing that I am a bit reluctant about is that `InputConfig` is not strongly typed, it is can contain any Key/Values, which can be error prone. Why don't we use the actual fields that are possible for filelog receiver?

UPDATE: I see you wrote this in the PR description:
>The stanza operators still do not make use of mapstructure, which forces the configuration to work with generic map[string]interface{} and map[interface{}]interface{}. This is messy and makes it difficult to write good tests. I want to clean this up soon.

Do you plan to address this in future PRs because it requires changes to log-collection library? If you plan to address in a future PR it is best to keep #2268 and #2282 open for now. (I took the liberty to put "Resolves" next to them in the PR description, but I was probably too quick, please revert back as needed).

File: receiver/filelogreceiver/filelog.go [hunk 3]
```diff
@@ -43,22 +42,45 @@ func (f ReceiverType) Type() configmodels.Type {
...
-			NameVal: typeStr,
+		BaseConfig: stanzareceiver.BaseConfig{
+			ReceiverSettings: configmodels.ReceiverSettings{
+				TypeVal: configmodels.Type(typeStr),
+				NameVal: typeStr,
+			},
+			Parsers: stanzareceiver.ParserConfigs{},
+		},
+		Input: stanzareceiver.InputConfig{
+			"type": "file_input",
```
Comment (tigrannajaryan) [inline receiver/filelogreceiver/filelog.go:57]: Why is this line needed?

File: receiver/filelogreceiver/filelog.go [hunk 4]
```diff
@@ -43,22 +42,45 @@ func (f ReceiverType) Type() configmodels.Type {
...
 // FileLogConfig defines configuration for the filelog receiver
 type FileLogConfig struct {
-	configmodels.ReceiverSettings `mapstructure:",squash"`
-	Operators                     stanzareceiver.OperatorConfig `mapstructure:"operators"`
+	stanzareceiver.BaseConfig `mapstructure:",squash"`
+	Input                     stanzareceiver.InputConfig `mapstructure:",remain"`
+}
+
+// DecodeInputConfig unmarshals the input operator
+func (f ReceiverType) DecodeInputConfig(cfg configmodels.Receiver) (*operator.Config, error) {
```
Comment (tigrannajaryan) [inline receiver/filelogreceiver/filelog.go:74]: The reason this exists is because  stanzareceiver.InputConfig is a generic map of any key/values, right? If we used strongly typed fields would this function become superfluous?

File: receiver/stanzareceiver/config.go [hunk 5]
```diff
@@ -15,26 +15,41 @@
...
+	"github.com/open-telemetry/opentelemetry-log-collection/operator"
+	"go.opentelemetry.io/collector/config/configmodels"
 	"gopkg.in/yaml.v2"
 )
 
-// OperatorConfig is an alias that allows for unmarshaling outside of mapstructure
+// BaseConfig is the common configuration of a stanza-based receiver
+type BaseConfig struct {
+	configmodels.ReceiverSettings `mapstructure:",squash"`
+	Parsers                       ParserConfigs `mapstructure:"parsers"`
```
Comment (tigrannajaryan) [inline receiver/stanzareceiver/config.go:26]: I am probably missing something. Is it not possible to declare `Parsers []operator.Config` here? Do we need the `decodeParserConfigs` function?

**Review by tigrannajaryan — COMMENTED** [review]
@djaglowski thanks a lot for making great progress on this. :-)

File: receiver/filelogreceiver/README.md (hunk 1 above)
Comment (djaglowski) [reply receiver/filelogreceiver/README.md:49]: You make a good point. Splitting operators into multiple buckets would make the order or operations ambiguous. It also would be difficult to code/maintain.

I think that `parsers` is a good term in the vast majority of cases, but perhaps it's best to leave this alone for now.

File: receiver/filelogreceiver/filelog.go (hunk 2 above)
Comment (djaglowski) [reply receiver/filelogreceiver/filelog.go:70]: As you noted, `everything that is not declared by stanzareceiver.BaseConfig is captured in Input`, which is not ideal, but I don't think this has any meaningful impact on how one would configure or use the receiver.

Although I had hoped to address the migration to mapstructure as part of this PR, I found the changes to run deeper than expected. 

As a result, I pivoted to this design with the intention of satisfying the user-facing requirements of the three tickets. I believe this PR gets us far enough that we can call those tickets complete and that refactoring for the sake of maintainability can proceed independently. Issue #2334 captures this work.

File: receiver/filelogreceiver/filelog.go (hunk 4 above)
Comment (djaglowski) [reply receiver/filelogreceiver/filelog.go:74]: Correct. I don't believe we'll need this method at all.

File: receiver/stanzareceiver/config.go (hunk 5 above)
Comment (djaglowski) [reply receiver/stanzareceiver/config.go:26]: The alias exists to provide an opportunity for us to use [operator.Config.UnmarshalYAML](https://github.com/open-telemetry/opentelemetry-log-collection/blob/4622fc64438e4f05767b60c0b5ac6441b8408302/operator/config.go#L70), which is called when `decodeParserConfigs` calls `yaml.Unmarshal`, because the type is called. 

This method contains the logic that unmarshals different types of operators to the associated configurations. In theory, I think what you are suggesting might work if we had an equivalent mapstructure decode hook.

**Review by tigrannajaryan — APPROVED** [review]
LGTM

**Comment by djaglowski** [comment]
@tigrannajaryan I believe this is ready to merge.
